blog/2018/12/12/dancer-and-minion/index.markdown

```diff
@@ -0,0 +1,299 @@
+    return @invalid_queues;
+}
+
+sub run_job( $self, $args ) {
+    my $job_name = $args->{ name     } or die "run_job(): must define job name!";
+    my $guid     = $args->{ guid     } or die "run_job(): must have GUID to process!";
+    my $title    = $args->{ title    } // $job_name;
+    my $queue    = $args->{ queue    } // 'default';
+    my $job_args = $args->{ job_args };
+
+    die "run_job(): Invalid job queue '$queue' specified" if $self->has_invalid_queues( $queue );
+
+    my %notes = ( title => $title, guid  => $guid );
+
+    return $self->runner->enqueue( $job_name => $job_args => { notes => \%notes, queue => $queue });
+}
+
+1;
+
+```
+## Creating Jobs
+
+In our base model class (Moose-based), we would create an attribute for our job runner:
+```
+has 'job_runner' => (
+    is      => 'ro',
+    isa     => 'Empowered::JobQueue',
+    lazy    => 1,
+    default => sub( $self ) { return Empowered::JobQueue->new->runner; },
+);
+```
+And in the models themselves, creating a new queueable task was as easy as:
+```
+$self->runner->add_task( InstantXML => sub( $job, $request_path, $guid, $company_db, $force, $die_on_error = 0 ) {
+    $job->note( 
+        request_path => $request_path,
+        feed_id      => 2098,
+        group        => $company_db,
+    );
+    MyJob::Models::FooBar->new( request_path => $request_path )->generate_xml({
+        pdf_guid     => $guid,
+        group        => $company_db,
+        force        => $force,
+        die_on_error => $die_on_error,
+    });
+});
+
+```
+
+## Running Jobs
+
+Starting a job from Dancer was super easy:
+
+```
+use Dancer2;
+use MyJob::JobQueue;
+
+sub job_queue {
+    return MyJob::JobQueue->new;
+}
+
+get '/my/api/route/:guid/:group/:force' => sub {
+    my $guid  = route_parameters->get( 'guid' );
+    my $group = route_parameters->get( 'group' );
+    my $force = route_parameters->get( 'force' );
+
+    debug "GENERATING XML ONLY FOR $guid";
+    job_queue->run_job({
+        name     => "InstantXML",
+        guid     => $guid,
+        title    => "Instant XML Generator",
+        queue    => 'InstantXML',
+        job_args => [ $self->request_path, $guid, $group, $force ],
+    }); 
+}
+```
+
+## Creating and Configuring the Job Queue Worker
+
+We wanted to easily configure our Minions for all hosts and environments in one spot. Since we use a lot of YAML in Dancer, specifying the Minion configuration in YAML made a lot of sense
+to us:
+
+```
+# What port does the dashboard listen on?
+dashboard_port: 4000
+
+# Add the rest later.
+dashboards:
+    UNKNOWN: http://localhost:3000/
+    DEV: http://my.development.host.tld:8001/
+
+# Hosts that have no entry assume the default configuration
+default:
+    max_children: 4
+    queues:
+        - default
+
+# Host-specific settings
+jcrome-precision-3510:
+    max_children: 8
+    queues:
+        - default
+        - InstantXML
+        - PayrollXML
+        - ChangeRequest
+```
+Our job queue workers look like:
+```
```

Comment:
This is, again, quite a long script without much explanation. Could you perhaps highlight a few key features before the script or else break it into parts?

Reply:
Done!